### ISSUE TITLE
Hierarchical Probabilistic Neural Network Language Model

### ISSUE THREAD

**Issue by eleveyuan** [eleveyuan/PR#6]
论文地址: http://proceedings.mlr.press/r5/morin05a/morin05a.pdf

**Comment by eleveyuan** [comment]
由于语言模型中训练数据的庞大，往往会给模型训练以及计算带来很大的工作量。word2vec为了提高其训练效率引入了Hierarchical Softmax这一方法来提高计算的效率

此篇讨论的是优化学习函数<img src="https://render.githubusercontent.com/render/math?math=f(w_t, w_{t-1}, ..., w_{t-n\oplus1}) = \frac{e^{-g(w_t, w_{t-1}, ..., w_{t-n \oplus1})}}{\sum _v e^{w_t, w_{t-1}, ..., w_{t-n\oplus1}}}">

对比Bengio Y前两篇论文网络结构发生了一些变化:
前两篇：
<img src="https://render.githubusercontent.com/render/math?math=y = b \oplus Wx \oplusUtanh(d \oplus Hx)">

此篇：
<img src="https://render.githubusercontent.com/render/math?math=y = b \oplus a^T Utanh(c \oplus Wx \oplus UF _v ^t)">

其中x为当前单词上文的单词，Fv为当前单词。

考虑的词与词之间有一定的相似性，相似的词语属于同一类。作者引入了**Hierarchical decomposition**。因为不同类的单词以及上下文进行预测时满足:
<img src="https://render.githubusercontent.com/render/math?math=P(Y=y|X=x)=P(Y=y|C=c(y),X)P(C=c(y)|X=x)">(c为cluster)
易证明有：
<img src="https://render.githubusercontent.com/render/math?math=P(Y|X)=\sum _i P(Y, C=i|X)=\sum _i P(Y|C=i, X)P(C=i|X)=P(Y|C=c(Y),X)P(C=c(Y)|C)">
这是因为上文单词x，下文单词只会与某个cluster中的Y有上下文关系

通过对单词与层级聚类进行一个bit vector编码，对于v有<img src="https://render.githubusercontent.com/render/math?math=(b_1(v), ...,b_m(v))">，如果<img src="https://render.githubusercontent.com/render/math?math=b_1(v)=1">表明单词v属于第一层的左子树，如果<img src="https://render.githubusercontent.com/render/math?math=b_2(v)=0">则表明v属于第二层右子树(假设左子树为1)。通过这样的表示来构建一个树(如果是一棵平衡树，将会大大的提速)

这样上下文依赖关系所表达的条件概率：
<img src="https://render.githubusercontent.com/render/math?math=P(v|w_{t-1}, ..., w_{t-n\oplus1})=\prod ^m _{j=1}P(b_j(v)|b_1(v), ...,b_{j-1}(v), w_{t-1}, ...,w_{t-n\oplus1})">
模型就变成了计算单词属于哪一类的概率，其中计算一个节点属于哪一类的方式：
<img src="https://render.githubusercontent.com/render/math?math=P(b|node, w_{t-1}, ..., w_{t-n\oplus1})=sigmoid(\alpha _{node} \oplus \beta ^T tanh(c \oplus Wx \oplusUN_{node}))">
其中每从树root到node一条路径都共享Nnode向量。最终整个模型预测每个单词的平均长度只有log|V|的长度。

**Comment by eleveyuan** [comment]
<img src="https://render.githubusercontent.com/render/math?math=\oplus">表示加号，github issue不能渲染加号,用这个符号暂时替代
